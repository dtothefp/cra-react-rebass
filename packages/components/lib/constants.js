export const WALMART = 'Walmart';
export const CVS = 'CVS';
export const WALGREENS = 'Walgreens';
export const TARGET = 'Target';
export const RITE_AID = 'Rite-Aid';
export const RALPHS = 'Ralphs';
export const VONS = 'Vons';
export const COSTCO = 'Costco';

export const chaingIds = {
  [WALMART]: 1,
  [CVS]: 2,
  [WALGREENS]: 3,
  [TARGET]: 6,
  [RITE_AID]: 23357,
  [RALPHS]: 20052,
  [VONS]: 93177,
  [COSTCO]: 31240,
};
