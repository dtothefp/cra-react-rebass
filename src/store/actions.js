import {
  ADD_FILTER,
  REMOVE_FILTER,
  SET_PHARMACIES
} from './constants';

export const setPharmacies = (pharmacies) => ({
  type: SET_PHARMACIES,
  pharmacies
});

export const addFilter = (value) => ({
  type: ADD_FILTER,
  value: value.toUpperCase(),
});

export const removeFilter = (value) => ({
  type: REMOVE_FILTER,
  value: value.toUpperCase(),
});
