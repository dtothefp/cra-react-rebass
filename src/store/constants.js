export const ADD_FILTER = `ADD_FILTER`;
export const REMOVE_FILTER = `REMOVE_FILTER`;
export const SET_PHARMACIES = `SET_PHARMACIES`;
