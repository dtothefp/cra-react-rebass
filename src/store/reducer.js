import {
  ADD_FILTER,
  REMOVE_FILTER,
  SET_PHARMACIES
} from './constants';

export default (state = {}, action) => {
  let updatedState;

  switch (action.type) {
    case SET_PHARMACIES:
      updatedState = {
        ...state,
        pharmacies: action.pharmacies
      };
      break;
    case ADD_FILTER:
      updatedState = {
        ...state,
        filter: [ ...state.filter, action.value ]
      };
      break;
    case REMOVE_FILTER:
     updatedState = {
        ...state,
        filter: state.filter.filter((value) => value !== action.value),
      };
      break;
    default:
      updatedState = state;
    break;
  }

  if (process.env.NODE_ENV === `development`) {
    console.log(`***UPDATED_STATE***`, updatedState);
  }

  return updatedState;
}
