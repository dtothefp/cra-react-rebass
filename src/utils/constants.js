export const WALMART = `WALMART`;
export const CVS = `CVS`;
export const WALGREENS = `WALGREENS`;
export const TARGET = `TARGET`;
export const RITE_AID = `RITE-AID`;
export const RALPHS = `RALPHS`;
export const VONS = `VONS`;
export const COSTCO = `COSTCO`;

export const chainIds = {
  [WALMART]: 1,
  [CVS]: 2,
  [WALGREENS]: 3,
  [TARGET]: 6,
  [RITE_AID]: 23357,
  [RALPHS]: 20052,
  [VONS]: 93177,
  [COSTCO]: 31240,
};
