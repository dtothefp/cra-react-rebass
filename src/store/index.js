import React, { createContext, useReducer } from 'react';
import PropTypes from 'prop-types';
import reducer from './reducer';
import * as _actions from './actions';
import * as _constants from './constants';

export const actions = _actions;
export const constants = _constants;

export const StoreContext = createContext();

const _Provider = ({children, store = {}}) => {
  const [ state, dispatch ] = useReducer(reducer, store);

  return (
    <StoreContext.Provider value={{state, dispatch}}>
      {children}
    </StoreContext.Provider>
  );
};

_Provider.propTypes = {
  store: PropTypes.object,
  children: PropTypes.element.isRequired
}

export const Provider = _Provider;
