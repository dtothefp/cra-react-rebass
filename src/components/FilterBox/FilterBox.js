import React, { useContext } from 'react';
import { Flex } from 'rebass';
import { Checkbox, Label } from '@rebass/forms';
import { chainIds } from '../../utils/constants';
import { actions, StoreContext } from '../../store';

const {addFilter, removeFilter} = actions;
const ids = Object.keys(chainIds);
const desktopWidth = 100 / ids.length;
const midWidth = desktopWidth * 2;

const FilterBox = () => {
  const {dispatch} = useContext(StoreContext);
  const handleChange = (e) => {
    const {checked, name} = e.target;

    dispatch(
      checked ? addFilter(name) : removeFilter(name)
    );
  };

  return (
    <Flex
      as="form"
      flexWrap='wrap'
    >
      {ids.map((name, i) => (
        <Label
          key={`${name}-${i}`}
          my={[1, 1, 0]}
          width={[`50%`, `${midWidth}%`, `${desktopWidth}%`]}
        >
          <Checkbox
            name={name.toLowerCase()}
            onChange={handleChange}
            sx={{
              position: `relative`,
              top: `-4px`,
              backgroundColor: `transparent !important`,
            }}
          />
          {name}
        </Label>
      ))}
    </Flex>
  );
};

export default FilterBox;
