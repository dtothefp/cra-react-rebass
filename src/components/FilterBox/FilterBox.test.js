import React from 'react';
import { fireEvent, queryByAttribute, render } from '@testing-library/react';
import FilterBox from './FilterBox';
import { actions, StoreContext } from '../../store';
import { WALMART } from '../../utils/constants';

describe(`#FilterBox`, () => {
  const dispatch = jest.fn();
  const state = {filter: []};
  const {addFilter, removeFilter} = actions;

  it(`renders`, () => {
    const {container} = render(
      <StoreContext.Provider value={{dispatch, state}}>
        <FilterBox />
      </StoreContext.Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it(`updates the state when checked`, () => {
    const {container} = render(
      <StoreContext.Provider value={{dispatch, state}}>
        <FilterBox />
      </StoreContext.Provider>
    );

    const name = WALMART.toLowerCase();
    const walmart = queryByAttribute(`name`, container, name);

    fireEvent.click(walmart);
    expect(dispatch.mock.calls[0][0]).toEqual(addFilter(name));

    fireEvent.click(walmart);
    expect(dispatch.mock.calls[1][0]).toEqual(removeFilter(name));
  });
});

