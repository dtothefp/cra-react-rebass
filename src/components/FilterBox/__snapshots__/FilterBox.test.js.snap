// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#FilterBox renders 1`] = `
<div>
  <form
    class="css-z54kij"
  >
    <label
      class="css-1dtl1qx"
    >
      <div
        class="css-vurnku"
      >
        <input
          class="css-p19g2b"
          name="walmart"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="css-1cbmnb9"
          fill="currentcolor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
        <svg
          aria-hidden="true"
          class="css-4cyfn"
          fill="currentcolor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </div>
      WALMART
    </label>
    <label
      class="css-1dtl1qx"
    >
      <div
        class="css-vurnku"
      >
        <input
          class="css-p19g2b"
          name="cvs"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="css-1cbmnb9"
          fill="currentcolor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
        <svg
          aria-hidden="true"
          class="css-4cyfn"
          fill="currentcolor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </div>
      CVS
    </label>
    <label
      class="css-1dtl1qx"
    >
      <div
        class="css-vurnku"
      >
        <input
          class="css-p19g2b"
          name="walgreens"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="css-1cbmnb9"
          fill="currentcolor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
        <svg
          aria-hidden="true"
          class="css-4cyfn"
          fill="currentcolor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </div>
      WALGREENS
    </label>
    <label
      class="css-1dtl1qx"
    >
      <div
        class="css-vurnku"
      >
        <input
          class="css-p19g2b"
          name="target"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="css-1cbmnb9"
          fill="currentcolor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
        <svg
          aria-hidden="true"
          class="css-4cyfn"
          fill="currentcolor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </div>
      TARGET
    </label>
    <label
      class="css-1dtl1qx"
    >
      <div
        class="css-vurnku"
      >
        <input
          class="css-p19g2b"
          name="rite-aid"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="css-1cbmnb9"
          fill="currentcolor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
        <svg
          aria-hidden="true"
          class="css-4cyfn"
          fill="currentcolor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </div>
      RITE-AID
    </label>
    <label
      class="css-1dtl1qx"
    >
      <div
        class="css-vurnku"
      >
        <input
          class="css-p19g2b"
          name="ralphs"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="css-1cbmnb9"
          fill="currentcolor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
        <svg
          aria-hidden="true"
          class="css-4cyfn"
          fill="currentcolor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </div>
      RALPHS
    </label>
    <label
      class="css-1dtl1qx"
    >
      <div
        class="css-vurnku"
      >
        <input
          class="css-p19g2b"
          name="vons"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="css-1cbmnb9"
          fill="currentcolor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
        <svg
          aria-hidden="true"
          class="css-4cyfn"
          fill="currentcolor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </div>
      VONS
    </label>
    <label
      class="css-1dtl1qx"
    >
      <div
        class="css-vurnku"
      >
        <input
          class="css-p19g2b"
          name="costco"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="css-1cbmnb9"
          fill="currentcolor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
        <svg
          aria-hidden="true"
          class="css-4cyfn"
          fill="currentcolor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </div>
      COSTCO
    </label>
  </form>
</div>
`;
